### PR TITLE
Fix bad conversion of block index using negative Y

### PR DESCRIPTION
**Pull request by AlexanderMandera**
Fixes issue #808.

We got the same problem and I debugged everything.
I found out that the NBT is sent right but the X/Y/Z coordinates inside the ChunkData packet is wrong:
image

The expected coordinates for this would be -72 -42 -10.

I found out that the block index does not respect negative values (I think because it was missed during 1.18 Minestom update). It did work pre-1.18 because there weren't any tile entities on Y < 0, right?

But now when there are negative integers, the two-complement value does not convert back from the block index the right way.

I am no expert at binary arithmetics but I tried
* to use the 28th bit of the index to make a negative flag
* use the other bits to store the absolute value of Y and
* check for the negative flag in the `getBlockPosition` function to negate the value again.

I have tested it and it works fine now (signs appear again on negative Y values).

I don't know if this is the best solution for this problem, so I appreciate some hints to do it better. :)

### CODE REVIEW

File: src/main/java/net/minestom/server/utils/chunk/ChunkUtils.java [hunk 1]
```diff
@@ -198,7 +198,12 @@ public static int getBlockIndex(int x, int y, int z) {
         z = z % Chunk.CHUNK_SIZE_Z;
 
         int index = x & 0xF; // 4 bits
```
Comment (Eoghanmc22) [inline src/main/java/net/minestom/server/utils/chunk/ChunkUtils.java:200]: I dont think this is needed

File: src/main/java/net/minestom/server/utils/chunk/ChunkUtils.java [hunk 2]
```diff
@@ -211,7 +216,9 @@ public static int getBlockIndex(int x, int y, int z) {
      */
     public static @NotNull Point getBlockPosition(int index, int chunkX, int chunkZ) {
         final int x = blockIndexToChunkPositionX(index) + Chunk.CHUNK_SIZE_X * chunkX;
-        final int y = index >>> 4 & 0xFF;
+        int y = index >>> 4 & 0xFF;
```
Comment (Eoghanmc22) [inline src/main/java/net/minestom/server/utils/chunk/ChunkUtils.java:219]: i think this should be `final int y = index << 4 >> 8 & 0xFF`

Comment (Eoghanmc22) [reply src/main/java/net/minestom/server/utils/chunk/ChunkUtils.java:219]: actually, why is this mask `& 0xFF` maybe this should use blockIndexToChunkPositionY and then change that to use the change i mentioned?

**Comment by TheMode** [comment]
Could you test it in `CoordinateTest`?

**Comment by AlexanderMandera** [comment]
@TheMode I found that the affected functions are not covered by any tests.
Everything in CoordinateTest passed though.

**Comment by TheMode** [comment]
Yes, you have to create new tests

**Comment by AlexanderMandera** [comment]
Ok, got it, will do it later

File: src/test/java/net/minestom/server/coordinate/CoordinateTest.java [hunk 3]
```diff
@@ -100,4 +100,18 @@ public void toSectionRelativeCoordinate() {
         assertEquals(0, ChunkUtils.toSectionRelativeCoordinate(32));
         assertEquals(1, ChunkUtils.toSectionRelativeCoordinate(33));
     }
+
+    @Test
+    public void testBlockIndex() {
```
Comment (TheMode) [inline src/test/java/net/minestom/server/coordinate/CoordinateTest.java:105]: You need to test every possibility (ensuring that all coordinates give a different index, prevent potential overflow). There is otherwise no way to know if you broke something.

There is also no need to use `Pos`

Comment (AlexanderMandera) [reply src/test/java/net/minestom/server/coordinate/CoordinateTest.java]: I have tried to get suitable test cases. I don't know if that are all but I also checked for Y overflow.

Y values between -2^23+1 to 2^23-1 convert correctly.
Everything above and below that results in a wrong conversion.

I don't know if I should catch this case and throw an exception, what do you think?
I think it is unlikely that someone wants and can use such high or low Y coordinates?

Comment (TheMode) [reply src/test/java/net/minestom/server/coordinate/CoordinateTest.java]: Seems good to me
